File: Windows_csharp/RoutingMessageQueue.cs
/*
 * Title:  spider RoutingMessageQueue.cs (Windows csharp)
 * Author: Shuichiro Endo
 */

using System;
using System.Collections.Generic;
using System.Threading;

namespace spider
{
    public class RoutingMessageQueue
    {
        private const int ROUTING_MESSAGE_QUEUE_CAPACITY = 100;
        private Queue<RoutingMessage> queue = new Queue<RoutingMessage>();
        private SemaphoreSlim token = new SemaphoreSlim(0, ROUTING_MESSAGE_QUEUE_CAPACITY);
        private SemaphoreSlim guard = new SemaphoreSlim(1, 1);
        private int count = 0;

        public RoutingMessageQueue()
        {

        }

        ~RoutingMessageQueue()
        {

        }

        public int GetCount()
        {
            return count;
        }

        public void Push(RoutingMessage message)
        {
            guard.Wait();
            try
            {
                while(token.CurrentCount >= ROUTING_MESSAGE_QUEUE_CAPACITY)
                {
                    guard.Release();
                    Thread.Sleep(10);
                    guard.Wait();
                }
                queue.Enqueue(message);
                count++;
                token.Release();
            }finally
            {
                guard.Release();
            }
        }

        public int PushTimeout(RoutingMessage message,
                               int tv_sec,
                               int tv_usec)
        {
            int ret = 0;
            int t = tv_sec * 1000 + tv_usec / 1000; // milliseconds

            if(guard.Wait(t))
            {
                try
                {
                    while(token.CurrentCount >= ROUTING_MESSAGE_QUEUE_CAPACITY)
                    {
                        guard.Release();
                        Thread.Sleep(10);
                        guard.Wait();
                    }
                    queue.Enqueue(message);
                    count++;
                    token.Release();
                }finally
                {
                    guard.Release();
                }
            }else
            {
#if DEBUGPRINT
                Console.WriteLine("[-] push timeout");
#endif
                ret = -1;
            }

            return ret;
        }

        public RoutingMessage Pop()
        {
            RoutingMessage message;

            token.Wait();
            guard.Wait();
            message = queue.Dequeue();
            count--;
            guard.Release();

            return message;
        }

        public RoutingMessage PopTimeout(int tv_sec,
                                         int tv_usec)
        {
            RoutingMessage message = null;
            int t = tv_sec * 1000 + tv_usec / 1000; // milliseconds

            if(token.Wait(t))
            {
                guard.Wait();
                message = queue.Dequeue();
                count--;
                guard.Release();
            }else
            {
#if DEBUGPRINT
                Console.WriteLine("[-] pop timeout");
#endif
            }

            return message;
        }

        public RoutingMessage PopLatestMessage()
        {
            RoutingMessage message = null;

            for(int i=count; i>0; i--)
            {
                message = this.Pop();
            }

            return message;
        }
    }
}

File: extra/Windows/routingmessagequeue.cpp
/*
 * Title:  spider routingmessagequeue.cpp (Windows)
 * Author: Shuichiro Endo
 */

#include "spider.hpp"
#include "routingmessagequeue.hpp"
#include "routingmessage.hpp"

namespace spider
{
    Routingmessagequeue::Routingmessagequeue()
    {

    }

    Routingmessagequeue::~Routingmessagequeue()
    {

    }

    int32_t Routingmessagequeue::get_count()
    {
        return this->count;
    }

    void Routingmessagequeue::push(std::shared_ptr<Routingmessage> message)
    {
        guard.acquire();
        while(this->count >= ROUTING_MESSAGE_QUEUE_CAPACITY)
        {
            guard.release();
            std::this_thread::sleep_for(std::chrono::milliseconds(10));
            guard.acquire();
        }
        queue.push(message);
        this->count++;
        token.release();
        guard.release();

        return;
    }

    int32_t Routingmessagequeue::push_timeout(std::shared_ptr<Routingmessage> message,
                                              int32_t tv_sec,
                                              int32_t tv_usec)
    {
        int32_t ret = 0;
        int64_t t = (int64_t)tv_sec * 1000000 + (int64_t)tv_usec;
        auto duration = std::chrono::microseconds{t};

        if(guard.try_acquire_for(duration))
        {
            while(this->count >= ROUTING_MESSAGE_QUEUE_CAPACITY)
            {
                guard.release();
                std::this_thread::sleep_for(std::chrono::milliseconds(10));
                guard.acquire();
            }
            queue.push(message);
            this->count++;
            token.release();
            guard.release();
        }else
        {
#ifdef DEBUGPRINT
            std::printf("[-] push timeout\n");
#endif
            ret = -1;
        }

        return ret;
    }

    std::shared_ptr<Routingmessage> Routingmessagequeue::pop()
    {
        token.acquire();
        guard.acquire();
        std::shared_ptr<Routingmessage> message = queue.front();
        queue.pop();
        this->count--;
        guard.release();

        return message;
    }

    std::shared_ptr<Routingmessage> Routingmessagequeue::pop_timeout(int32_t tv_sec,
                                                                     int32_t tv_usec)
    {
        std::shared_ptr<Routingmessage> message = nullptr;
        int64_t t = (int64_t)tv_sec * 1000000 + (int64_t)tv_usec;
        auto duration = std::chrono::microseconds{t};

        if(token.try_acquire_for(duration))
        {
            guard.acquire();
            message = queue.front();
            queue.pop();
            this->count--;
            guard.release();
        }else
        {
#ifdef DEBUGPRINT
            std::printf("[-] pop timeout\n");
#endif
        }

        return message;
    }

    std::shared_ptr<Routingmessage> Routingmessagequeue::pop_latest_message()
    {
        std::shared_ptr<Routingmessage> message = nullptr;

        for(int32_t i=this->count; i>0; i--)
        {
            message = this->pop();
        }

        return message;
    }
}

